mantidproject/mantid
# Fix Workbench shutdown RuntimeError and GlobalFigureManager tests

**Description of work.**
Fixes the RuntimeError _sometimes_ shown on Windows on shutdown.

<details>
<summary>Some details for the error cause</summary>
The reason was the unhandled `DictionaryAction.Clear` was being sent from the Figure's ObservableDictionary to the Figure Manager. This would default to updating all figures. This update would be triggered on Workbench shutdown. 
The error showing was a race condition between whether Qt will free the widget first, or the plots will be updated first.

`DictionaryAction.Clear` is now properly handled and triggers `destroy` on all live figures for proper disposal.

</details>
<br>

A small refactoring was done in `globalfiguremanager.py`.

Tests have been added for the `GlobalFigureManager` and `GlobalFigureManagerObserver` -> `workbench.test_globalfiguremanager.test_globalfiguremanager`

**To test:**
Run `ctest` with `-R workbench` and check that all Workbench tests pass.
Open Workbench and plot a few things, try out all the options


Fixes #24050




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
